62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+
+                        if (isMonotonic)
+                        {
+                            // Already was monotonic
+                            addBezier(unsplitBezier);
+                            if (debugOutput)
+                                drawDebugBezier(unsplitBezier, float32_t4(0.8, 0.8, 0.8, 0.2));
+                        }
+                        else
+                        {
+                            addBezier(monotonicSegments.data()[0]);
+                            addBezier(monotonicSegments.data()[1]);
+                            if (debugOutput)
+                            {
+                                drawDebugBezier(monotonicSegments.data()[0], float32_t4(0.0, 0.6, 0.0, 0.5));
+                                drawDebugBezier(monotonicSegments.data()[1], float32_t4(0.0, 0.0, 0.6, 0.5));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        for (uint32_t bezierIdx = 0; bezierIdx < beziers.size(); bezierIdx++)
+        {
+            auto hatchBezier = &beziers[bezierIdx];
+            Segment segment;
+            segment.originalBezier = hatchBezier;
+            segment.t_start = 0.0;
+            segment.t_end = 1.0;
+            segments.push_back(segment);
+        }
+
+        // TODO better way to do this
+        std::sort(segments.begin(), segments.end(), [&](Segment a, Segment b) { return a.originalBezier->p[0][major] > b.originalBezier->p[0][major]; });
+		for (Segment& segment : segments)
+		{
+			starts.push(segment);
+			//std::cout << "Starts: " << segment.originalBezier->p[0][major]  << "\n";
+		}
+
+        std::sort(segments.begin(), segments.end(), [&](Segment a, Segment b) { return a.originalBezier->p[2][major] > b.originalBezier->p[2][major]; });
+        for (Segment& segment : segments)
+		{
+			ends.push(segment.originalBezier->p[2][major]);
+			//std::cout << "Ends: " << segment.originalBezier->p[2][major] << "\n";
+		}
+        maxMajor = segments.front().originalBezier->p[2][major];
+    }
+
+    // Sweep line algorithm
+	std::priority_queue<double, std::vector<double>, std::greater<double> > intersections; // Next intersection points as major coordinate
+    std::vector<Segment> activeCandidates; // Set of active candidates for neighbor search in sweep line
+
+	// if we weren't spawning quads, we could just have unsorted `vector<Bezier*>`
+	auto candidateComparator = [&](const Segment& lhs, const Segment& rhs)
+	{
+		// btw you probably want the beziers in Quadratic At^2+B+C form, not control points
+		double _lhs = lhs.originalBezier->evaluateBezier(lhs.t_start)[minor];
+		double _rhs = rhs.originalBezier->evaluateBezier(rhs.t_start)[minor];
+
+		auto len = [](float64_t2 vec)
+		{
+			return (double)sqrt(vec.x * vec.x + vec.y * vec.y);
+		};
+		double lenLhs = len(lhs.originalBezier->p[2] - lhs.originalBezier->p[0]);
+		double lenRhs = len(rhs.originalBezier->p[2] - rhs.originalBezier->p[0]);
```

Comment:
you can use glm::distance or glm::length, we don't need to do these basic math operations ourselves now that we're using glm on cpp side